Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.mm

```diff
@@ -237,15 +238,24 @@
     });
 }
 
-void PDFDataDetectorOverlayController::didInvalidateHighlightOverlayRects(ShouldUpdatePlatformHighlightData shouldUpdatePlatformHighlightData, ActiveHighlightChanged activeHighlightChanged)
+void PDFDataDetectorOverlayController::didInvalidateHighlightOverlayRects(std::optional<PDFDocumentLayout::PageIndex> pageIndex, ShouldUpdatePlatformHighlightData shouldUpdatePlatformHighlightData, ActiveHighlightChanged activeHighlightChanged)
 {
     // Regardless of what we repaint, we don't need the stale data after this.
     auto resetStaleDataDetectorWithHighlight = makeScopeExit([&] {
         m_staleDataDetectorItemWithHighlight = { { }, { } };
     });
 
     if (shouldUpdatePlatformHighlightData == ShouldUpdatePlatformHighlightData::Yes) {
-        WTF::forEach(m_pdfDataDetectorItemsWithHighlightsMap.keys(), [&](auto pageIndex) {
+        auto pageIndices = [&] {
+            if (pageIndex) {
+                if (auto iterator = m_pdfDataDetectorItemsWithHighlightsMap.find(*pageIndex); iterator != m_pdfDataDetectorItemsWithHighlightsMap.end())
+                    return makeSizedIteratorRange(m_pdfDataDetectorItemsWithHighlightsMap, iterator.keys(), std::next(iterator).keys());
+            }
+
+            return m_pdfDataDetectorItemsWithHighlightsMap.keys();
+        }();
+
+        WTF::forEach(pageIndices, [this](auto pageIndex) {
```

Comment:
why can't this be a normal `for x : pageIndices` loop?

Reply:
No real reason, I'd just written it that way. I can simplify it to `for x : pageIndices`

Reply:
(Will do so in an upcoming patch...)